### ISSUE TITLE
MSVC 15.9.0 Preview ICE in cast.h

### ISSUE THREAD

**Issue by patrikhuber** [pybind/pybind11#1551]
Hi,

Looks like Microsoft may have introduced another regression into their compiler. VS 15.9.0-pre.3.0 (MSVC 19.16.26926.0) throws an ICE in `cast.h`:

> eos\3rdparty\pybind11\include\pybind11\cast.h(1594): fatal error C1001: An internal error has occurred in the compiler.

The line in question is

`std::is_same<decltype(std::declval<make_caster<T>>().operator T&()), T&>`

which is part of the expression

```
template <typename T> struct move_if_unreferenced<T, enable_if_t<all_of<
    move_is_plain_type<T>,
    negation<move_always<T>>,
    std::is_move_constructible<T>,
    std::is_same<decltype(std::declval<make_caster<T>>().operator T&()), T&>
>::value>> : std::true_type {};
```

I've reported this to MS: https://developercommunity.visualstudio.com/content/problem/347743/ice-with-pybind11.html

I know you're usually not fixing bugs in Preview but I thought it be a good idea to report & track this problem here too, in case others stumble upon it, or in case MS does not fix this for 15.9 Final.

**Comment by patrikhuber** [comment]
There has been progress in the above-linked developercommunity report. MS fixed this internally and it will be part of one of the next preview releases. So the really good news is that this bug will **not** be in 15.9 Final. I think it's not worth putting a workaround into pybind11, as it only affects a Preview version. I would suggest to leave this issue here open until 15.9 Final is released and then close it, just so that if anyone else encounters this issue right now, they'll find this open issue (if they do a search).

**Comment by CdTCzech** [comment]
Fixed in Visual Studio 2017 version 15.9 Preview 5.

**Comment by YannickJadoul** [comment]
Thanks for reporting, reporting to MS, and for following up with the solution, @patrikhuber & @CdTCzech!

Closing.
